### PR TITLE
Make the plist backend pluggable

### PR DESCRIPTION
**Pull request by segiddins**
- [x] Documentation
- [x] CHANGELOG
- [x] Tests

\c @alloy @neonichu 

My brain was fried so I whipped this up

### CODE REVIEW

**Comment by alloy** [comment]
Does it occur often that a user can't use the FFI version or what else is the motivation?

**Comment by segiddins** [comment]
@alloy it means you can safely require Xcodeproj on, say, Linux, and be OK as long as you don't need the ascii plist stuff (basically just delaying the point of failure)

**Comment by neonichu** [comment]
Looks fine to me :+1: — I am assuming the code in `lib/xcodeproj/plist/ffi/*` is just copy-pasted? I did not review that in detail for changes.

**Comment by alloy** [comment]
Ok, looks good to me to too, the code is definitely much nicer laid out :+1:

**Comment by segiddins** [comment]
Yeah, just copy-pasted. I do want the stuff in plist.rb reviewed, though. It's super gross and still needs tests.

File: Gemfile.lock [hunk 1]
```diff
@@ -93,4 +93,4 @@ DEPENDENCIES
   xcodeproj!
 
 BUNDLED WITH
-   1.10.6
+   1.11.0.pre.2
```
Comment (segiddins) [inline Gemfile.lock]: revert this bit

File: lib/xcodeproj/plist/plist_gem.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  module Plist
+    module PlistGem
+      def self.available?
+        @available ||= begin
+          require 'plist/parser'
+          require 'plist/generator'
+          true
+        rescue LoadError
+          'Xcodeproj relies on a library called `plist` to read and write ' \
+          'Xcode project files. Ensure yo have the `plist` gem installed and ' \
```
Comment (segiddins) [inline lib/xcodeproj/plist/plist_gem.rb]: `you`

File: lib/xcodeproj/plist.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    end
+
+    # @return [Bool] Checks whether there are merge conflicts in the file.
+    #
+    # @param  [#to_s] path
+    #         The path of the file.
+    #
+    def self.file_in_conflict?(path)
+      file = File.open(path, 'r')
+      file.each_line.any? { |l| l.match(/^(<|=|>){7}/) }
```
Comment (segiddins) [inline lib/xcodeproj/plist.rb]: this could just be `File.read().match()`

File: lib/xcodeproj/plist.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+      unless path.is_a?(String) || path.is_a?(Pathname)
+        raise TypeError, "The given `#{path}` must be a string or 'pathname'."
+      end
+      path = path.to_s
+      raise IOError, 'Empty path.' if path.empty?
+      implementation.write_to_path(hash, path)
+    end
+
+    POSSIBLE_IMPLEMENTATIONS = [:FFI, :PlistGem]
```
Comment (segiddins) [inline lib/xcodeproj/plist.rb]: `KNOWN_IMPLEMENTATIONS`

File: lib/xcodeproj/plist.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      def implementation
+        @implementation ||= autoload_implementation
+      end
+    end
+
+    def self.autoload_implementation
+      failures = POSSIBLE_IMPLEMENTATIONS.map do |impl|
+        begin
+          impl = Plist.const_get(impl)
+          available = impl.available?
```
Comment (segiddins) [inline lib/xcodeproj/plist.rb]: rename `available?` to something more intuitive

**Comment by segiddins** [comment]
This should now be ready to merge, please give it a final review?

**Comment by segiddins** [comment]
:shipit: since I got no comments

**Pull request merged**